qt/src/hocr/HOCRPdfExporter.cc

```diff
@@ -779,3 +659,211 @@ void HOCRPdfExporter::updateValid() {
+                    PoDoFo::PdfPage* pdfpage = document->CreatePage(PoDoFo::PdfRect(0, 0, pageWidth, pageHeight));
+                    double offsetX = 0.5 * (pageWidth - bbox.width() * docScale);
+                    double offsetY = 0.5 * (pageHeight - bbox.height() * docScale);
+                    pdfprinter.setPage(pdfpage, docScale, offsetX, offsetY);
+                    printChildren(pdfprinter, page, pdfSettings, imgScale);
+                    if(pdfSettings.overlay) {
+                        QRect scaledBBox(imgScale * bbox.left(), imgScale * bbox.top(), imgScale * bbox.width(), imgScale * bbox.height());
+                        QImage selection;
+                        QMetaObject::invokeMethod(this, "getSelection",  Qt::BlockingQueuedConnection, Q_RETURN_ARG(QImage, selection), Q_ARG(QRect, scaledBBox));
+                        pdfprinter.drawImage(bbox, selection, pdfSettings);
+                    }
+                    QMetaObject::invokeMethod(this, "setSource", Qt::BlockingQueuedConnection, Q_RETURN_ARG(bool, success), Q_ARG(QString, page->sourceFile()), Q_ARG(int, page->pageNr()), Q_ARG(int, sourceDpi), Q_ARG(double, page->angle()));
+                    painter.FinishPage();
+                } else {
+                    failed.append(page->title());
+                }
+            }
+            monitor.increaseProgress();
+        }
+        return true;
+    }, _("Exporting to PDF..."));
+    MAIN->hideProgress();
+    if(!failed.isEmpty()) {
+        QMessageBox::warning(MAIN, _("Errors occurred"), _("The following pages could not be rendered:\n%1").arg(failed.join("\n")));
+    }
+
+    // Set PDF info
+    PoDoFo::PdfInfo* pdfInfo = document->GetInfo();
+    pdfInfo->SetProducer(ui.lineEditProducer->text().toStdString());
+    pdfInfo->SetCreator(ui.lineEditCreator->text().toStdString());
+    pdfInfo->SetTitle(ui.lineEditTitle->text().toStdString());
+    pdfInfo->SetSubject(ui.lineEditSubject->text().toStdString());
+    pdfInfo->SetKeywords(ui.lineEditKeywords->text().toStdString());
+    pdfInfo->SetAuthor(ui.lineEditAuthor->text().toStdString());
+
+    QString errMsg;
+    bool success = pdfprinter.finalize(&errMsg);
+    if(!success) {
+        QMessageBox::warning(MAIN, _("Export failed"), _("The PDF export failed (%1).").arg(errMsg));
+    }
+    return success;
+}
+
+bool HOCRPdfExporter::exportePdfByQt(const QString& outname, const QPageSize &pageSize, const PDFSettings &pdfSettings)
+{
+    //init printer
+    QPrinter printer;
+    printer.setOutputFormat(QPrinter::PdfFormat);
+    printer.setOutputFileName(outname);
+    printer.setResolution(m_hocrdocument->page(0)->resolution());
+    printer.setPageSize(pageSize);
+    printer.setCreator(ui.lineEditCreator->text());
+    printer.setFullPage(true);
+
+    QFont defaultFont = ui.checkBoxFontFamily->isChecked() ? ui.comboBoxFontFamily->currentFont() : ui.comboBoxFallbackFontFamily->currentFont();
+
```

Comment:
The code after this part should be pretty much printer-agnostic. Can't you keep common for both QPrinter and PoDoFo, and just have dedicated functions for the painter initialization?

Reply:
Hi, What does dedicated functions mean?  It means same as special function?

Reply:
Do you mean that I should add more error-check code in **exporterbyQt** function?

Reply:
No, I mean that you should be able to leave the entire block of code below in `HOCRPdfExporter::run`, and just add something like `HOCRPdfExporter::createQPrinter` and `HOCRPdfExporter::createPoDoFoPrinter` which does the backend specific initialization of the printers.

Reply:
I got your idea

Reply:
Thanks  for your advice~